### PR TITLE
4nec2 Antenna Calculation

### LINKED ISSUE #29: SEM52SL Antenna
We need to create a binarized antenna for the SEM52SL. As a placeholder it uses a 120cm BNC.

It should use either a 40cm or 90cm BNC.

### LINKED ISSUE #201: Switch antenna data models from MMANA-GAL to 4nec2
**Description:**
- At present the binary gain data for the anntenas is generated from MMANA-GAL, switching to 4nec2 would make things easier for others to produce these models and the data format is easier to work with and could provide better simulation information.

**Tasks:**
- [x] Update C++ signal processing to use the new format
- [x] Generate data for the currently used antennas
- [x] Create a tutorial.

### PR DESCRIPTION
**Pull request by Soldia1138**
**When merged this pull request will:**
- Port all antenna data to generation by 4NEC2
- Add missing antennas 
- Change C++ Code to handle data
- Close #201 
- Close #29

### CODE REVIEW

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+---
+title: Adding a new antenna
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: Use `Adding New Antenna` and rename file to `adding-new-antenna.md` for consistency with other pages.

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+---
+
+Every radio in ACRE can be equipped with at least one antenna. Since radios usually have specific antennas designed and tuned for the radio and the frequency range used, it is likely one must add a new antenna if a [new radio](http://acre.idi-systems.com/wiki/frameworks/creating-new-radio.html) is created for ACRE.
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: - `ACRE2`
- `[new radio](creating-new-radio)`

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+In order to allow an antenna to work properly in ACRE, there are a few required config entries and most importantly a set of gain data covering the desired frequency range. The gain data is used by the signal processing functions for calculation of signal strengths. For basics refer to [Radio Loss](http://acre2.idi-systems.com/wiki/user/radio-signal-loss.html).
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: - `ACRE2`
- `[Radio Loss](radio-signal-loss)`

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+## Example Antenna Config
+
+Antennas themselves are defined in the `sys_antennas` component of ACRE2 [sys_antennas\CfgAcreComponents.hpp](https://github.com/IDI-Systems/acre2/blob/master/addons/sys_antenna/CfgAcreComponents.hpp). The parameters should be self explanatory as the following examples shows.
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: Forward slash in path.

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        polarization = VERTICAL_POLARIZE;
+        heightAG = AVERAGE_MAN_HEIGHT;
+        orient = 90;    // in degrees off of flat plane
+        name = "Default Antenna";
+        connector = ACRE_CONNECTOR_TNC;
+        height = 1.2; //meters
+        binaryGainFile = "\idi\acre\addons\sys_antenna\binary\VHF_1.2m_whip_gain.aba";
+    };
+
+    class ACRE_2HALFINCH_UHF_TNC : ACRE_BaseAntenna {
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: No space before `:`

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+```
+class CfgAcreComponents {
+    class ACRE_ComponentBase;
+    
+    class ACRE_BaseAntenna: ACRE_ComponentBase {
+        type = ACRE_COMPONENT_ANTENNA;
+        simple = true;
+        polarization = VERTICAL_POLARIZE;
+        heightAG = AVERAGE_MAN_HEIGHT;
+        orient = 90;    // in degrees off of flat plane
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: One space before comment.

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        name = "Default Antenna";
+        connector = ACRE_CONNECTOR_TNC;
+        height = 1.2; //meters
+        binaryGainFile = "\idi\acre\addons\sys_antenna\binary\VHF_1.2m_whip_gain.aba";
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: Can just use `QPATHTOF(binary\<antenna>.aba)`

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    };
+
+    class ACRE_2HALFINCH_UHF_TNC : ACRE_BaseAntenna {
+        name = "2.5 Inch UHF Antenna AN/PRC-343 ONLY";
+        connector = ACRE_CONNECTOR_TNC;
+        height = 0.062457; //meters
+        binaryGainFile = "\idi\acre\addons\sys_antenna\binary\prc343_gain.aba";
+    };
+};
+
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: No empty line.

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+```
+
+## Generating Antenna Gain Data
+
+As ACRE tries to simulate radios as close to reality as possible, the gain data used is coming from a tool used for antenna and electromagnetic effects simulation. This tool is called 4NEC2 and can be obtained for free: [4NEC2](http://www.qsl.net/4nec2/?)
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: - `ACRE2`
- No question mark at end of URL.

File: docs/wiki/frameworks/adding-a-new-antenna.md [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+For installation instructions and a bunch of good tutorials, please refer to the website. 
+
+### Setting up the antenna
+
+The first step after opening 4NEC2 is to create a model of the antenna. Usually it is more than sufficient to use a vertical wire and a simple ground setting. These models are stored in the _.nec_ format which is basically a text file. The _.nec_ files from all existing antennas could be found in the `extras/antenna` folder and a good practice is to simply take an existing file and change the parameters either in a text editor or directly in 4NEC2.
```
Comment (jonpas) [inline docs/wiki/frameworks/adding-a-new-antenna.md]: `.nec` in code ticks.

**Review by jonpas — CHANGES_REQUESTED** [review]
Linsk to other pages in same directory can be shoretened to just the file name without extension.

File: addons/sys_antenna/CfgAcreComponents.hpp [hunk 11]
```diff
@@ -23,18 +23,18 @@ class CfgAcreComponents {
...
     
-    class ACRE_120CM_VHF_TNC: ACRE_BaseAntenna {
-        name = "1.2 Meter VHF Antenna TNC";
+    class ACRE_100CM_VHF_TNC : ACRE_BaseAntenna {
+        name = "1 Meter VHF Antenna TNC";
         connector = ACRE_CONNECTOR_TNC;
-        height = 1.2; //meters
-        binaryGainFile = "\idi\acre\addons\sys_antenna\binary\VHF_1.2m_whip_gain.aba";
+        height = 1; //meters
+        binaryGainFile = "\idi\acre\addons\sys_antenna\binary\Thales_100cm_Whip_gain.aba";
```
Comment (jonpas) [inline addons/sys_antenna/CfgAcreComponents.hpp]: `QPATHTOF(binary\Thales_100cm_Whip_gain.aba)` in all.

File: addons/sys_antenna/CfgAcreComponents.hpp [hunk 12]
```diff
@@ -23,18 +23,18 @@ class CfgAcreComponents {
         binaryGainFile = "\idi\acre\addons\sys_antenna\binary\UHF_14_inch_gain.aba";
     };
     
-    class ACRE_120CM_VHF_TNC: ACRE_BaseAntenna {
-        name = "1.2 Meter VHF Antenna TNC";
+    class ACRE_100CM_VHF_TNC : ACRE_BaseAntenna {
```
Comment (jonpas) [inline addons/sys_antenna/CfgAcreComponents.hpp]: You like spaces? 💃

File: addons/sys_antenna/CfgAcreComponents.hpp [hunk 13]
```diff
@@ -57,4 +57,32 @@ class CfgAcreComponents {
         height = 0.062457; //meters
         binaryGainFile = "\idi\acre\addons\sys_antenna\binary\oe303_gain.aba";
     };
+
+    class ACRE_SEM52_SHORT_BNC : ACRE_BaseAntenna {
+        name            = "SEM52SL Antenna Short (0.4m)";
```
Comment (jonpas) [inline addons/sys_antenna/CfgAcreComponents.hpp]: Just one space on each side of `=`.

File: extensions/src/ACRE2Arma/signal/antenna/antenna.cpp [hunk 14]
```diff
@@ -30,7 +33,7 @@ float acre::signal::antenna::gain(const glm::vec3 dir_antenna_, const glm::vec3
 {
     if (f_ < _min_frequency || f_ + _frequency_step > _max_frequency)
         return -1000.0f;
-
+
```
Comment (jonpas) [inline extensions/src/ACRE2Arma/signal/antenna/antenna.cpp]: Whitespace.

File: tools/antenna_create_binary.py [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+import struct
+
+classname = sys.argv[1]
+
+maindirectory = os.path.abspath(os.path.join(os.path.dirname( __file__ ), '..'))
+filepath = os.path.join(maindirectory,'extras','antennas')
+
+###################################
+#
+#	Read Values from Output file
```
Comment (jonpas) [inline tools/antenna_create_binary.py]: Tabs (whole file).

File: extras/antennas/AT-271.inp [hunk 16]
```diff
@@ -0,0 +1,17 @@
+CM PRC343 Antenna
```
Comment (jonpas) [inline extras/antennas/AT-271.inp]: Sure? 🤔

File: extensions/src/ACRE2Arma/signal/antenna/antenna.cpp [hunk 17]
```diff
@@ -39,8 +42,14 @@ float acre::signal::antenna::gain(const glm::vec3 dir_antenna_, const glm::vec3
     float elev_signal = asin(dir_signal_v.z)*57.2957795f;
     float dir_signal = atan2(dir_signal_v.x, dir_signal_v.y)*57.2957795f;
 
+	LOG(TRACE) << "Elevation Antenna : " << elev_antenna;
```
Comment (jonpas) [inline extensions/src/ACRE2Arma/signal/antenna/antenna.cpp]: Tabs (whole file).

**Review by jonpas — CHANGES_REQUESTED** [review]
Taaabs.

File: docs/wiki/frameworks/adding-a-new-antenna.md (hunk 3 above)
Comment (Soldia1138) [reply docs/wiki/frameworks/adding-a-new-antenna.md]: But that isn't in the same directory, is it?

Comment (jonpas) [reply docs/wiki/frameworks/adding-a-new-antenna.md]: Right, my bad, should be `[Radio Loss](/wiki/user/radio-signal-loss)` then.

**Review by jonpas — DISMISSED** [review]

**Comment by jonpas** [comment]
Added #29 reference to OP as this PR adds the SEM52SL antenna.

**Review by jonpas — APPROVED** [review]

**Pull request merged**